File: test/test_clock.py
#
# Tests for functions in clock.py
#
# Copyright (c) 2015 WatSat-ADCS
# Licensed under the MIT license.
#
# authors: Jason Pye (j2pye@uwaterloo.ca)
#
# Change log:
# 2015-10-02 (JP) - Initial release
#

import unittest
import clock

class testClock(unittest.TestCase):

  def test_readClock(self):
    self.assertIsNotNone( clock.readClock() )

  def test_writeClock(self):
    new_time = 1234.567890
    clock.writeClock(new_time)
    self.assertEqual(clock.readClock(), new_time)

  def test_resetClock(self):
    clock.resetClock()
    self.assertEqual(clock.readClock(), 2451725.03494)

  def test_tickClock(self):
    dt = 1
    init_time = clock.readClock()
    clock.tickClock(dt)
    self.assertEqual(clock.readClock(), init_time + dt)

  def test_greg2jdate(self):
    greg_year = 2000
    greg_month = 1
    greg_day = 1
    greg_hour = 12
    greg_minute = 00
    greg_second = 00
    jdate = 2451545.0
    self.assertEqual( '%.5f' % clock.greg2jdate(greg_year,greg_month,greg_day,greg_hour,greg_minute,greg_second), '%.5f' % jdate)

  def test_jdate2greg(self):
    greg_year = 2000
    greg_month = 1
    greg_day = 1
    greg_hour = 12
    greg_minute = 00
    greg_second = 00
    jdate = 2451545.0

    converted_greg = clock.jdate2greg(jdate)
    self.assertEqual(converted_greg['year'], greg_year)
    self.assertEqual(converted_greg['month'], greg_month)
    self.assertEqual(converted_greg['day'], greg_day)
    self.assertEqual(converted_greg['hour'], greg_hour)
    self.assertEqual(converted_greg['minute'], greg_minute)
    self.assertEqual(converted_greg['second'], greg_second)


if __name__ == '__main__':
  unittest.main()

File: lib/tests/test_clock.c
/*
  Tests for functions in clock.c

  Copyright (c) 2015 WatSat-ADCS
  Licensed under the MIT license.

  authors: Jason Pye (j2pye@uwaterloo.ca)

  Change log:
  2015-12-02 (JP) - Initial release
*/

#include <check.h>
#include <stdio.h>
#include <stdlib.h>
#include "clock.h"

const char clock_filename[] = "../../dat/jdate.txt";

/*===== TESTS =====*/

START_TEST (test_readClock) {

  double jdate = -1;
  char str_jdate[100];

  jdate = readClock(clock_filename);
  sprintf( str_jdate, "%.5lf", jdate );

  ck_assert( jdate > 0 );
  ck_assert_str_ne( str_jdate, "" );

} END_TEST


START_TEST (test_writeClock) {

  double new_time = 12345.12345;
  writeClock(clock_filename, new_time);
  ck_assert( readClock(clock_filename) == new_time );

} END_TEST


START_TEST (test_resetClock) {

  double DEFAULT_TIME = 2451725.03494;
  resetClock(clock_filename);
  ck_assert( readClock(clock_filename) == DEFAULT_TIME );

} END_TEST


START_TEST (test_tickClock) {

  double dt = 1.;
  double init_time = readClock(clock_filename);
  tickClock(clock_filename, dt);
  ck_assert( readClock(clock_filename) == init_time + dt );

} END_TEST




/*===== SUITE =====*/

#define SUITE_NAME "Clock"

Suite * create_suite() {

  Suite *s = suite_create(SUITE_NAME);

  TCase *tc_core = tcase_create("Core");
  tcase_add_test(tc_core, test_readClock);
  tcase_add_test(tc_core, test_writeClock);
  tcase_add_test(tc_core, test_resetClock);
  tcase_add_test(tc_core, test_tickClock);
  suite_add_tcase(s, tc_core);

  return s;
}


int main() {

  int num_fail;
  Suite *s = create_suite();
  SRunner *sr = srunner_create(s);
  srunner_run_all(sr, CK_NORMAL);
  num_fail = srunner_ntests_failed(sr);
  srunner_free(sr);

  return (num_fail == 0) ? EXIT_SUCCESS : EXIT_FAILURE;

}
